### DISCUSSION TITLE
How to call WinApi functions from dbgeng sources?

### DISCUSSION THREAD

**Discussion by MrROBUST** [NationalSecurityAgency/ghidra#4913] [Q&A]
Hello, I trying to debug a Windows application. When a debugger starts it gives a warning:

> WARN  Launch th12.exe - .ProgramDB: The resulting target process has no mapping to the static image th12.exe - .ProgramDB. Intervention is required before static and dynamic addresses can be translated. Check the target's module list.   (ConsoleErrorDisplay.java:56)

AFAIK this happens because Ghidra cant match a source execubable file name with loaded module

WinDbg names this module like: `consoleOutput: ModLoad: 00000000`00400000 00000000`004d9000   image00000000`00400000`

The file `Ghidra/Debug/Debugger/src/main/java/ghidra/app/plugin/core/debug/service/model/launch/AbstractDebuggerProgramLaunchOffer.java` has line 88:
```
else if (info.type == PageType.MAPPED) {
    // TODO: Figure out the file name
}
```

Probably there we could just call a [GetModuleFileNameExW](https://learn.microsoft.com/en-us/windows/win32/api/psapi/nf-psapi-getmodulefilenameexw), to get actual file name. But what class the Ghidra use to call WinApi?

**Comment by d-millar** [comment]
@MrROBUST Hi and thanks for your question.  The short answer is, **with th12.exe in the Static Listing View and the debugger running, right-click in the Modules View on the line containing the executable that matches your static program image and select "Map Module to th12.exe".**

Why does this error occur?  Well, unfortunately, "name" is a bit of an ambiguous concept here.  The code can be compiled with one name, potentially saved on disk under another, loaded into Ghidra under another, and renamed at any point either in Ghidra or on disk.  We need to match the "program"'s name (in Ghidra's sense of the term) with the load name during execution (which typically matches the name at the time of compilation).  Technically, we could search the .rsrc section and probably make that match, but doing so from inside the debugger is a bit complicated.  

We have been working to improve this code.  The code you referenced, which I believe is in DbgListMemoryRegionsCommand.java, isn't really relevant.  Also, the WinApi isn't particularly useful in this case.  We use the dbgeng (and less frequently the dbghelp) API to retrieve the executing module names, but they are relatively unambiguous.  That is to say, mismatches between the compilation-time name and the execution-time name are infrequent, except in the case of malware.  The problem is really recovering the compilation-time name from the staic executable, aka the Ghidra program.

Let us know if the initial suggestion doesn't work for you for some reason, however, and/or if you have further suggestions.  Always appreciated!

D

**Comment by MrROBUST** [comment]
@d-millar Thank you for reply.

I still think something is missing in the current implementation of the debugger. Using WinAPI, we can write a simple program to display all loaded process modules (btw, I found that Ghidra uses JNA to access WinAPI):
<details>
  <summary>Expand the code</summary>

```C++
#include <iostream>
#include <string_view>
#include <vector>

#include <windows.h>
#include <psapi.h>

void oops()
{
  auto err = GetLastError();
  __debugbreak();
}

int main()
{
  DWORD pid = 13112;

  auto hProcess = OpenProcess(PROCESS_QUERY_LIMITED_INFORMATION | PROCESS_VM_READ, FALSE, pid);
  if (!hProcess)
  {
    oops();
  }

  DWORD lpcbNeeded = 0;
  {
    // just to get required size
    HMODULE dummy[1];
    if (!EnumProcessModulesEx(hProcess, dummy, sizeof(dummy), &lpcbNeeded, LIST_MODULES_ALL))
    {
      oops();
    }
  }
  std::vector<HMODULE> lphModule;
  lphModule.resize(lpcbNeeded / sizeof(HMODULE));
  if (!EnumProcessModulesEx(hProcess, lphModule.data(), lpcbNeeded, &lpcbNeeded, LIST_MODULES_ALL))
  {
    oops();
  }

  for (auto hModule : lphModule)
  {
    WCHAR modulePath[MAX_PATH];
    auto size = GetModuleFileNameEx(hProcess, hModule, modulePath, MAX_PATH);
    if (size == 0)
    {
      oops();
      continue;
    }

    MODULEINFO modinfo;
    if (!GetModuleInformation(hProcess, hModule, &modinfo, sizeof(MODULEINFO)))
    {
      oops();
      continue;
    }

    std::string fileName;
    {
      // just to get required size
      size = WideCharToMultiByte(CP_UTF8, WC_ERR_INVALID_CHARS, modulePath, size, fileName.data(), 0, 0, FALSE);
      if (size == 0)
      {
        oops();
        continue;
      }
      fileName.resize(size);
    }
    size = WideCharToMultiByte(CP_UTF8, WC_ERR_INVALID_CHARS, modulePath, size, fileName.data(), size, 0, FALSE);
    if (size == 0)
    {
      oops();
      continue;
    }

    std::cout << modinfo.EntryPoint << " "
              << modinfo.SizeOfImage << " "
              << modinfo.lpBaseOfDll << " "
              << fileName << std::endl;
  }

  CloseHandle(hProcess);

  return 0;
}
```

</details>
The results are quite similar to what the process information visualizers show. We can get the file path/name and base address.
<details>
  <summary>Image</summary>

![modules](https://user-images.githubusercontent.com/5786945/214102150-066c80e6-04e0-4ce4-b79a-7766c1d60cea.png)

</details>

Perhaps the malware can use some tricks to hide from this solution, but if we can fix a simple case, why not give it a try?

I'll look to WinDbg COM interfaces if it can query module information directly (Plan B is to use WinAPI).

Could you please give some advice on where to look for the relevant debugger implementation?

**Comment by d-millar** [comment]
@MrROBUST Ghidra uses JNA to access the dbgeng.dll API.  I would be a little hesitant to call that the WinAPI, but....

If you look at DbgListModulesCommand::invoke, you can follow the call sequence down to the JNA call in DebugSymbolsImpl1::callNamesForWhich, which invokes IDebugSymbol::GetModuleName.  That said, this is not the issue.  This information is generally available and more often than not matches the compilation-time name.

The problem is that that name, i.e. the name retrieved dynamically through the debugger API, may not match the name of the static program imported into Ghidra, which may or may not be derived from its name on disk which may or may not be related to its compilation-time name.  There is currently no Ghidra API for retrieving the compilation-time name from the program AFAIK.

Sorry, if I'm being unclear - the topic is a little confusing generally.

**Comment by d-millar** [comment]
@MrROBUST   Following up again:  I just looked a little more closely at you images....

Is the name of your executable "th12.exe" everywhere, i.e. in the Program, in the "About" dialog, in the Modules View?  
If so, I have misunderstood your problem - I think the real problem is that you have spaces in the path and code in charge of testing for a match is dropping the rest of the line.

**Reply by MrROBUST** [reply]
Removed spaces from path. Timeout happens and module path is not recognized, named as image00000000`00400000
Process Hacker queries it correctly.

<details>
  <summary>Image</summary>

![modules](https://user-images.githubusercontent.com/5786945/214222839-5b019f88-63c7-4fe7-8cc1-bb6bfee01e3b.png)

</details>

Makeshift example from above fails on first EnumProcessModulesEx with GetLastError == 299 (0x12B ERROR_PARTIAL_COPY), probably because lacks some permissions.

**Comment by d-millar** [comment]
Hmmm, well I tested out my blanks-in-the-path-theory and, nope, that won't cause the problem.  The only other possibilities I can see are (1) System.getProperty("os.name") is returning somethin without 'windows' in the text, or (2) you are timing out for some other reason and the error message is masking the cause.  Do you get this error for all executables?  I.e. can you run "notepad.exe"?

**Comment by MrROBUST** [comment]
Running on Windows 10 x64, dbgeng via GADP.

notepad.exe works without timeouts

<details>
  <summary>Debugger output</summary>
  
  ```
No .natvis files found at C:\WINDOWS\SYSTEM32\Visualizers.
No .natvis files found at C:\Users\MrROBUST\AppData\Local\Dbg\Visualizers.

Microsoft (R) Windows Debugger Version 10.0.19041.1503 AMD64
Copyright (c) Microsoft Corporation. All rights reserved.

CommandLine: C:\Windows\System32\notepad.exe
Unable to add extension DLL: ntsdexts
Unable to add extension DLL: uext
Unable to add extension DLL: exts
SECURE: File not allowed to be loaded - C:\WINDOWS\SYSTEM32\dbghelp.dll
Error code: Win32 error 0n5
The call to LoadLibrary(ext) failed, Win32 error 0n2
    "Не удается найти указанный файл."
Please check your debugger configuration and/or network access.

************* Path validation summary **************
Response                         Time (ms)     Location
Deferred                                       srv*
Symbol search path is: srv*
Executable search path is: 
ModLoad: 00007ff7`afed0000 00007ff7`aff08000   notepad.exe
0:000>g
ModLoad: 00007fff`22db0000 00007fff`22fa8000   ntdll.dll
ModLoad: 00007fff`214f0000 00007fff`215af000   C:\WINDOWS\System32\KERNEL32.DLL
ModLoad: 00007fff`20a50000 00007fff`20d22000   C:\WINDOWS\System32\KERNELBASE.dll
ModLoad: 00007fff`214c0000 00007fff`214eb000   C:\WINDOWS\System32\GDI32.dll
ModLoad: 00007fff`20820000 00007fff`20842000   C:\WINDOWS\System32\win32u.dll
ModLoad: 00007fff`20550000 00007fff`2065f000   C:\WINDOWS\System32\gdi32full.dll
ModLoad: 00007fff`20850000 00007fff`208ed000   C:\WINDOWS\System32\msvcp_win.dll
ModLoad: 00007fff`20660000 00007fff`20760000   C:\WINDOWS\System32\ucrtbase.dll
ModLoad: 00007fff`21f70000 00007fff`22111000   C:\WINDOWS\System32\USER32.dll
ModLoad: 00007fff`216d0000 00007fff`21a25000   C:\WINDOWS\System32\combase.dll
ModLoad: 00007fff`21cd0000 00007fff`21df5000   C:\WINDOWS\System32\RPCRT4.dll
ModLoad: 00007fff`22290000 00007fff`2233d000   C:\WINDOWS\System32\shcore.dll
ModLoad: 00007fff`223a0000 00007fff`2243e000   C:\WINDOWS\System32\msvcrt.dll
....
  ```
  
</details>

A simple 64 bit console application works fine. As an example, I used the program from my previous answer.
A 32-bit build of the same program times out listenForMapping from AbstractDebuggerProgramLaunchOffer.java, but the module name is recognized.

<details>
  <summary>Debugger output</summary>
  
  ```
No .natvis files found at C:\WINDOWS\SYSTEM32\Visualizers.
No .natvis files found at C:\Users\MrROBUST\AppData\Local\Dbg\Visualizers.

Microsoft (R) Windows Debugger Version 10.0.19041.1503 AMD64
Copyright (c) Microsoft Corporation. All rights reserved.

CommandLine: C:\Users\MrROBUST\Desktop\MyTree\Release\helloworld-32.exe
Unable to add extension DLL: ntsdexts
Unable to add extension DLL: uext
Unable to add extension DLL: exts
SECURE: File not allowed to be loaded - C:\WINDOWS\SYSTEM32\dbghelp.dll
Error code: Win32 error 0n5
The call to LoadLibrary(ext) failed, Win32 error 0n2
    "Не удается найти указанный файл."
Please check your debugger configuration and/or network access.
Unable to add extension DLL: wow64exts

************* Path validation summary **************
Response                         Time (ms)     Location
Deferred                                       srv*
Symbol search path is: srv*
Executable search path is: 
ModLoad: 00000000`00340000 00000000`00348000   helloworld.exe
0:000>g
ModLoad: 00007fff`22db0000 00007fff`22fa8000   ntdll.dll
ModLoad: 00000000`77450000 00000000`775f4000   ntdll.dll
ModLoad: 00007fff`21a30000 00007fff`21a89000   C:\WINDOWS\System32\wow64.dll
ModLoad: 00007fff`21ed0000 00007fff`21f53000   C:\WINDOWS\System32\wow64win.dll
ModLoad: 00000000`77440000 00000000`7744a000   C:\WINDOWS\System32\wow64cpu.dll
ModLoad: 00000000`76b60000 00000000`76c50000   C:\WINDOWS\SysWOW64\KERNEL32.DLL
ModLoad: 00000000`76ce0000 00000000`76efc000   C:\WINDOWS\SysWOW64\KERNELBASE.dll
ModLoad: 00000000`707a0000 00000000`70840000   C:\WINDOWS\SysWOW64\apphelp.dll
ModLoad: 00000000`759e0000 00000000`75b00000   C:\WINDOWS\SysWOW64\ucrtbase.dll
ModLoad: 00000000`72bc0000 00000000`72c2d000   C:\WINDOWS\SysWOW64\MSVCP140.dll
ModLoad: 00000000`76820000 00000000`76903000   C:\WINDOWS\SysWOW64\ole32.dll
ModLoad: 00000000`762e0000 00000000`7639e000   C:\WINDOWS\SysWOW64\RPCRT4.dll
ModLoad: 00000000`75e90000 00000000`76110000   C:\WINDOWS\SysWOW64\combase.dll
ModLoad: 00000000`76a80000 00000000`76aa3000   C:\WINDOWS\SysWOW64\GDI32.dll
ModLoad: 00000000`75b80000 00000000`75b98000   C:\WINDOWS\SysWOW64\win32u.dll
ModLoad: 00000000`769a0000 00000000`76a80000   C:\WINDOWS\SysWOW64\gdi32full.dll
ModLoad: 00000000`75c60000 00000000`75cdb000   C:\WINDOWS\SysWOW64\msvcp_win.dll
ModLoad: 00000000`76580000 00000000`7671d000   C:\WINDOWS\SysWOW64\USER32.dll
ModLoad: 00000000`74990000 00000000`749a5000   C:\WINDOWS\SysWOW64\VCRUNTIME140.dll
ModLoad: 75c30000 75c56000   C:\WINDOWS\SysWOW64\IMM32.DLL
(2034.4040): WOW64 breakpoint - code 4000001f (first chance)
First chance exceptions are reported before any exception handling.
This exception may be expected and handled.
*** WARNING: Unable to verify checksum for helloworld.exe
0:000:x86>g
(2034.4040): WOW64 breakpoint - code 4000001f (first chance)
First chance exceptions are reported before any exception handling.
This exception may be expected and handled.
```

</details>

Is there something wrong with WoW64 causing a timeout? But I think module path recognition has a different problem.

**Comment by d-millar** [comment]
Hmmm, OK - well, getting more interesting!  Did you hit the timeout in your 32-bit helloworld.exe test?  I am not generally seeing a timeout on 32-bit processes (am testing with winmine.exe, right now).  I don't see any real errors in your listings above.  The dbghelp.dll load error occurs when the debugger tries to load dbghelp.dll from System32, which is no longer allowed on current Windows versions.  If you place a copy anywhere in the load path ahead of System32, that should go away.  Not sure what causes the checksum error, but I don't think that's significant.  The error message is a system message, not a Ghidra one.

Try running with dbgeng.dll IN-VM (versus GADP), maybe.  Having a separate agent, while smart from a crash perspective adds a layer of complexity that may be hiding the underlying error.

**Comment by d-millar** [comment]
OK, am going back to names as a possible reason for the timeout...   Looking at the images above, it see "th12.exe" in a bunch of places and "th12-.exe" in a bunch of places.   Is it possible the executable was compiled with one of those variations and loaded into Ghidra with another?  I realize this is a pretty small difference, but I don't think the mapper would have figured out that those two are a match.

**Reply by MrROBUST** [reply]
I'll check IN-VM debug in a moment.
Previously, I just made a copy of the application directory to make sure spaces weren't a problem. th12.exe has been renamed to th12-.exe to distinguish it from the source directory executable because I put them in the same Ghidra project. Sorry if this causes confusion, these files are the same.

**Comment by d-millar** [comment]
Understood re same, but was the Ghidra program you were debugging when you hit the timeout NAMED "th12-.exe" or did this happen for both "th12.exe" and "th12-exe"?

**Reply by MrROBUST** [reply]
I probably don't understand what name you mean.

In the file system and task managers, they are referred to as executable files: th12.exe and th12-.exe.

The debugger (and hence Ghidra) calls them both `image00000000'00400000`

Double checked this in IN-VM and in GADB.

**Comment by MrROBUST** [comment]
Checking IN-VM debug.
First, I was expecting the debugger to use the dbgeng.dll from the WinDbg installation directory (it's in the system PATH `C:\Program Files (x86)\Windows Kits\10\Debuggers\x64\`, but I saw some issues on github about how Java might not respect this). Well, the debugger works despite the error, right? Looks like it can still load the DLLs from system32 for some reason!

<details>
  <summary>Image</summary>
  
![modules](https://user-images.githubusercontent.com/5786945/214335013-d7ab1edc-c373-4e75-82b2-80ee272e3789.png)
  
</details>

- The notepad.exe just works.

- On a 64-bit console application, the debugger starts without problems. But when I type 'g' in the interpreter to continue execution, Ghidra gets an uncaught exception:

<details>
  <summary>Exception</summary>

java.lang.AssertionError
	at ghidra.app.plugin.core.debug.gui.listing.DebuggerListingProvider.doSetProgram(DebuggerListingProvider.java:594)
	at ghidra.app.plugin.core.debug.gui.listing.DebuggerListingProvider.lambda$goTo$3(DebuggerListingProvider.java:768)
	at utilities.util.SuppressableCallback.invokeWithTop(SuppressableCallback.java:186)
	at ghidra.app.plugin.core.debug.gui.listing.DebuggerListingProvider.goTo(DebuggerListingProvider.java:756)
	at ghidra.app.plugin.core.debug.gui.listing.DebuggerListingProvider.lambda$doGoToTracked$6(DebuggerListingProvider.java:1017)
	at java.desktop/java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:318)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:771)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:722)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:716)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:86)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:741)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)

---------------------------------------------------
Build Date: 2022-Nov-15 1249 EST
Ghidra Version: 10.2.2
Java Home: C:\Program Files\Eclipse Adoptium\jdk-17.0.6.10-hotspot
JVM Version: Eclipse Adoptium 17.0.6
OS: Windows 10 10.0 amd64
Workstation: DESKTOP-096SHCF

</details>

The application itself was not rebuilt to change the hardcoded PID to something valid, so the application was expected to crash or break when __debugbreak() is called. But something happend in Ghidra.
But then it turned out to be not so trouble. Ghidra and the debugger continued to run, so when I hit 'g' a few more times, the trace hit the breakpoint and the application stopped as expected.

<details>
  <summary>Debugger output</summary>

CommandLine: C:\Users\MrROBUST\Desktop\MyTree\x64\Release\helloworld.exe
Unable to add extension DLL: ntsdexts
Unable to add extension DLL: uext
Unable to add extension DLL: exts
SECURE: File not allowed to be loaded - C:\WINDOWS\SYSTEM32\dbghelp.dll
Error code: Win32 error 0n5
The call to LoadLibrary(ext) failed, Win32 error 0n2
    "Не удается найти указанный файл."
Please check your debugger configuration and/or network access.

************* Path validation summary **************
Response                         Time (ms)     Location
Deferred                                       srv*
Symbol search path is: srv*
Executable search path is: 
ModLoad: 00007ff6`3d620000 00007ff6`3d629000   helloworld.exe
0:000>g
ModLoad: 00007fff`22db0000 00007fff`22fa8000   ntdll.dll
ModLoad: 00007fff`214f0000 00007fff`215af000   C:\WINDOWS\System32\KERNEL32.DLL
ModLoad: 00007fff`20a50000 00007fff`20d22000   C:\WINDOWS\System32\KERNELBASE.dll

************* Path validation summary **************
Response                         Time (ms)     Location
Deferred                                       srv*
Symbol search path is: srv*
Executable search path is: 
ModLoad: 00007ff6`a5a80000 00007ff6`a5b5a000   conhost.exe
0:000>g
ModLoad: 00007fff`22db0000 00007fff`22fa8000   ntdll.dll
ModLoad: 00007fff`214f0000 00007fff`215af000   C:\WINDOWS\System32\KERNEL32.DLL
ModLoad: 00007fff`20a50000 00007fff`20d22000   C:\WINDOWS\System32\KERNELBASE.dll
ModLoad: 00007fff`20850000 00007fff`208ed000   C:\WINDOWS\System32\msvcp_win.dll
ModLoad: 00007fff`20660000 00007fff`20760000   C:\WINDOWS\System32\ucrtbase.dll
ModLoad: 00007fff`22290000 00007fff`2233d000   C:\WINDOWS\System32\shcore.dll
ModLoad: 00007fff`223a0000 00007fff`2243e000   C:\WINDOWS\System32\msvcrt.dll
ModLoad: 00007fff`216d0000 00007fff`21a25000   C:\WINDOWS\System32\combase.dll
ModLoad: 00007fff`21cd0000 00007fff`21df5000   C:\WINDOWS\System32\RPCRT4.dll
ModLoad: 00007fff`22b90000 00007fff`22c3e000   C:\WINDOWS\System32\advapi32.dll
ModLoad: 00007fff`21420000 00007fff`214bc000   C:\WINDOWS\System32\sechost.dll
ModLoad: 00007fff`21f70000 00007fff`22111000   C:\WINDOWS\System32\user32.dll
ModLoad: 00007fff`20820000 00007fff`20842000   C:\WINDOWS\System32\win32u.dll
ModLoad: 00007fff`214c0000 00007fff`214eb000   C:\WINDOWS\System32\GDI32.dll
ModLoad: 00007fff`20550000 00007fff`2065f000   C:\WINDOWS\System32\gdi32full.dll
ModLoad: 00007fff`22120000 00007fff`22152000   C:\WINDOWS\System32\IMM32.DLL
ModLoad: 00007fff`22440000 00007fff`22b84000   C:\WINDOWS\System32\shell32.dll
ModLoad: 00007fff`1de70000 00007fff`1df0e000   C:\WINDOWS\system32\uxtheme.dll
ModLoad: 00007fff`215b0000 00007fff`216c5000   C:\WINDOWS\System32\MSCTF.dll
ModLoad: 00007fff`21e00000 00007fff`21ecd000   C:\WINDOWS\System32\OLEAUT32.dll
ModLoad: 00007fff`0ee10000 00007fff`0eebc000   C:\WINDOWS\SYSTEM32\TextShaping.dll
ModLoad: 00007fff`0d990000 00007fff`0dc2a000   C:\WINDOWS\WinSxS\amd64_microsoft.windows.common-controls_6595b64144ccf1df_6.0.19041.1110_none_60b5254171f9507e\comctl32.DLL
ModLoad: 00007fff`1e050000 00007fff`1e07f000   C:\WINDOWS\SYSTEM32\dwmapi.dll
ModLoad: 00007fff`1e340000 00007fff`1e352000   C:\WINDOWS\SYSTEM32\kernel.appcore.dll
ModLoad: 00007fff`20760000 00007fff`207e2000   C:\WINDOWS\System32\bcryptPrimitives.dll
ModLoad: 00007fff`17430000 00007fff`17529000   C:\WINDOWS\SYSTEM32\textinputframework.dll
ModLoad: 00007fff`1d970000 00007fff`1da62000   C:\WINDOWS\System32\CoreMessaging.dll
ModLoad: 00007fff`1d610000 00007fff`1d96e000   C:\WINDOWS\System32\CoreUIComponents.dll
ModLoad: 00007fff`21a90000 00007fff`21afb000   C:\WINDOWS\System32\WS2_32.dll
ModLoad: 00007fff`1cf40000 00007fff`1d094000   C:\WINDOWS\SYSTEM32\wintypes.dll
ModLoad: 0000021e`400e0000 0000021e`40234000   C:\WINDOWS\SYSTEM32\wintypes.dll
ModLoad: 00007fff`1f580000 00007fff`1f5b3000   C:\WINDOWS\SYSTEM32\ntmarta.dll
ModLoad: 00007fff`22160000 00007fff`2220f000   C:\WINDOWS\System32\clbcatq.dll
ModLoad: 00007fff`20660000 00007fff`20760000   C:\WINDOWS\System32\ucrtbase.dll
ModLoad: 00007fff`01bd0000 00007fff`01c5e000   C:\WINDOWS\SYSTEM32\MSVCP140.dll
ModLoad: 00007fff`22c40000 00007fff`22d6a000   C:\WINDOWS\System32\ole32.dll
ModLoad: 00007fff`21cd0000 00007fff`21df5000   C:\WINDOWS\System32\RPCRT4.dll
ModLoad: 00007fff`216d0000 00007fff`21a25000   C:\WINDOWS\System32\combase.dll
ModLoad: 00007fff`214c0000 00007fff`214eb000   C:\WINDOWS\System32\GDI32.dll
ModLoad: 00007fff`20820000 00007fff`20842000   C:\WINDOWS\System32\win32u.dll
ModLoad: 00007fff`20550000 00007fff`2065f000   C:\WINDOWS\System32\gdi32full.dll
ModLoad: 00007fff`20850000 00007fff`208ed000   C:\WINDOWS\System32\msvcp_win.dll
ModLoad: 00007fff`21f70000 00007fff`22111000   C:\WINDOWS\System32\USER32.dll
ModLoad: 00007fff`1a6d0000 00007fff`1a6dc000   C:\WINDOWS\SYSTEM32\VCRUNTIME140_1.dll
ModLoad: 00007fff`0de40000 00007fff`0de5b000   C:\WINDOWS\SYSTEM32\VCRUNTIME140.dll
ModLoad: 00007fff`22120000 00007fff`22152000   C:\WINDOWS\System32\IMM32.DLL
(4404.2570): Break instruction exception - code 80000003 (first chance)
*** WARNING: Unable to verify checksum for helloworld.exe
0:000>g
(4404.2570): Break instruction exception - code 80000003 (first chance)
0:000>g
ModLoad: 00007fff`22c40000 00007fff`22d6a000   C:\WINDOWS\System32\ole32.dll
(4404.2570): Break instruction exception - code 80000003 (first chance)
0:000>g
ModLoad: 00007fff`1e340000 00007fff`1e352000   C:\WINDOWS\SYSTEM32\kernel.appcore.dll
ModLoad: 00007fff`223a0000 00007fff`2243e000   C:\WINDOWS\System32\msvcrt.dll


</details>

</details>

- On a 32-bit console application, a mapping timeout occurs, but the module is recognized by the debugger and by Ghidra:

`ModLoad: 00000000'00d30000 00000000'00d38000   helloworld.exe`

<html>
<body>


7ff63d620000 | 7ff63d628fff | helloworld.exe | helloworld.exe | [1‥+∞) | 36864
-- | -- | -- | -- | -- | --



</body>
</html>

The following trace continues without exception, hits the breakpoint, and exits.

- th12.exe: mapping timeout, module name is wrong

```
Executable search path is: 
ModLoad: 00000000'00400000 00000000'004d9000   image00000000'00400000
```

<html>
<body>


00400000 | 004d8fff | image00000000`00400000 | image00000000`00400000 | [1‥+∞) | 888832
-- | -- | -- | -- | -- | --



</body>
</html>

**Comment by d-millar** [comment]
OK, to summarize, I think I need to be looking into at least 3 problems here:

(1) what triggered that "doSetProgram" exception and how to fix it?
(2) what would cause you're test executable's name to show up as image00000000`00400000? 
(3) what accounts for the difference in behavior between the GADP and IN-VM variants?

(Arguably, I should also revisit the load path issue, but I've revisited that several times without making much progress.)

**Reply by MrROBUST** [reply]
I would also try to check. Can you tell me where is the relevant code to query and match module names?
Also is there an option for developers to disable the timeout? They messing with debugging. Right now I'm just changing `getTimeoutMillis()` to wait for an hour, but still...

Also relevant issues: https://github.com/NationalSecurityAgency/ghidra/issues/4732 https://github.com/NationalSecurityAgency/ghidra/issues/4583

**Comment by d-millar** [comment]
If you look in DebuggerStaticMappingProposals::namesContain, am pretty sure that's the code that tries to match the module(Lower)Name to a program, so, if you breakpoint there, you will at least be able to tell what the two names are and maybe why it's confused.  No current option to disable the timeout, although we have a previous ticket in for that, I believe.

**Comment by d-millar** [comment]
OK, googling around, it appears that the dbgeng API returns the "image0000000`00xxxxxxx" form for some (non-VS?) WOW64 processes.  I guess, returning to your original post, it might be an interesting question as to whether the WinAPI would return a different answer at the same point in the target's execution.  That said, am a little loathe to load the WinAPI for only this purpose.

**Reply by MrROBUST** [reply]
Well, querying the module filename with IDebugSymbols3::GetModuleNameString from dbgeng only returns "image*bla-bla-bla" for some reason.

If we can't use the GetModuleFileNameEx function from psapi, the only option I can find is the [EnumerateLoadedModules*](https://learn.microsoft.com/en-us/windows/win32/api/dbghelp/nf-dbghelp-enumerateloadedmodulesw64) functions from dbghelp. But this looks even less convenient and only returns the filename and a few wow64 dlls:
```
4194304 888832 E:\games\touhou\games\12 Undefined Fantastic Object\th12.exe
140733778165760 2064384 C:\WINDOWS\SYSTEM32\ntdll.dll
140733757718528 364544 C:\WINDOWS\System32\wow64.dll
140733762568192 536576 C:\WINDOWS\System32\wow64win.dll
2000945152 40960 C:\WINDOWS\System32\wow64cpu.dll
```

So in order to match a loaded 32-bit module with its filename in a Wow64 environment, the only reliable option I can think of is to match its base address with a list of all loaded modules.

**Reply by d-millar** [reply]
Except that's exactly why were doing this, i.e. we're trying to adjust the base address of the program to match the executing module.  These addresses will almost never match.
